src/PluginClientBuilder.php

```diff
@@ -0,0 +1,168 @@
+<?php
+
+namespace Http\Client\Common;
+
+use Http\Client\HttpClient;
+
+/**
+ * A builder that help you build a PluginClient.
+ *
+ * @author Tobias Nyholm <tobias.nyholm@gmail.com>
+ */
+class PluginClientBuilder implements PluginClientBuilderInterface
+{
+    /**
+     * @var Plugin[]
+     */
+    protected $plugins = [];
+
+    /**
+     * @var bool
+     */
+    protected $rebuildClient = true;
+
+    /**
+     * The last created client with the plugins
+     *
+     * @var PluginClient
+     */
+    private $pluginClient;
+
+    /**
+     * The object that sends HTTP messages
+     *
+     * @var HttpClient
+     */
+    private $httpClient;
+
+    /**
+     *
+     * @param HttpClient $httpClient
+     */
+    public function __construct(HttpClient $httpClient = null)
+    {
+        $this->httpClient = $httpClient;
+    }
+
+    /**
+     * Add a new plugin to the end of the plugin chain.
+     *
+     * @param Plugin $plugin
+     */
+    public function addPlugin(Plugin $plugin)
+    {
+        $this->plugins[] = $plugin;
+        $this->rebuildClient = true;
+    }
+
+    /**
+     * Remove a plugin by its fully qualified class name (FQCN).
+     *
+     * @param string $fqcn
+     */
+    public function removePlugin($fqcn)
+    {
+        foreach ($this->plugins as $idx => $plugin) {
+            if ($plugin instanceof $fqcn) {
+                unset($this->plugins[$idx]);
+                $this->rebuildClient = true;
+            }
+        }
+    }
+
+    /**
+     * Alias for removePlugin and addPlugin
+     *
+     * @param Plugin $plugin
+     */
+    public function replacePlugin(Plugin $plugin)
+    {
+        $this->removePlugin(get_class($plugin));
```

Comment:
maybe we should not have this method but have people get the plugins, change the list and set plugins. RequestMatcherPlugin for example is quite likely to be used multiple times.